docs/proposals/03-loadbalancer.md

```diff
@@ -0,0 +1,128 @@
+
+### Goals
+- define a user facing API (onMetal api) for the load balancer
+- enable the user to request an L3 Loadbalancer (only IP addresses) with load balancer address (outside facing) and 
+- with multiple target's based on the onMetal ([OEP-1](01-networking-integration.md)) network interface architecture
+- the targets of the load balancer must be dynamic changeable
+- the load balancer needs to avoid forwarding unintended traffic.
+  - based on ports/protocols (UDP/TCP/SCTP) it must filter what needs to be load balanced
+  - ICMP requests are not part of forwarded traffic and will be filtered by the load balancer anyhow (the loa)
+- Load balancer (IP) and targets can be parts of different networks but
+- all targets (NetworkInterfaces) must be of the same network 
+
+### Non-Goals
+- no address or port translation / rewriting (no SNAT / DNAT) (L4 Loadbalancer) support
+- no injection of additional information (e.g. x-forwarded-for) (L7 Loadbalancer) support
+- no protocol offloading like ssl (L7 Loadbalancer) support
+- if multiple load balancer IP's are needed, request multiple load balancers (a load balancer services one load balancer IP)
+
+### Details
+- load balancing is used to deliver a packet addressed to the load balancer to one of its targets via the onMetal network routing
+- the load balancing will be for the target and the source transparent
+- the target needs to know the load balancers IP and needs to answer with it (and to receive traffic with it)
+- answers to request will be directly delivered since all details are known to the target
+- payload packages are not changed, to not lose any information
+
+## Proposal
+A network loadbalancer CRD allows the user to define the network function. all entries except the `targetNetworkInterface` must be immutable. 
+
+### User defined object
+```yaml
+apiVersion: networking.api.onmetal.de/v1alpha1
+kind: LoadBalancer
+metadata: 
+  name: myLoadBalancer-2abf34
+  namespace: customer-1
+spec:
+  #the load balancer IP. is a generated IP and only of type public (internet IP) 
+  #TODO other types of IP addresses (e.g. internal LB) and pinned IP types
+  type: Public
+  #a load balancer can have two IP addresses: IPv4 or IPv6 but not more addresses IPfamily
+  ipFamilies: [ IPv4, IPv6 ]
+
+  #filtering the load balanced traffic by protocol and port/port ranges 
+  ports:
+    - name: webserver  # might be optional
+      # protocols supported UDP, TCP, SCTP
+      protocol: tcp
+      # single port
+      port: 80
+    - name: db
+      protocol: udp
+      # portrange
+      port: 1024
+      portEnd: 2048
+  #all networkInterfaces must be connected to this network.
+  networkRef:
+    name:
+  #a selector for the NetworkInterfaces to be load balancer targets, normal kubernetes selector logic
+  #TODO explicit machines ?
+  networkInterfaceSelector:
+    matchLabels: 
+      key: db
+      foo: bar
+status:
+  #the ips the load balancer is listening to (max 2, an IPv4 and/or IPv6)
```

Comment:
For extensible scalability it'd be good to support multiple IPs per load balancer, i.e. supply and configure them with same set of rules.

Then, it'd be possible to put them into the RRDNS and have multiple resolvable IP addresses per endpoint.

Reply:
That's exactly the non-goal we mentioned - if multiple IPs should be supported, multiple LBs have to be created.

Reply:
in theory that is a nice idea, but the idea so so nice scalable that there should be no technical reason for having multiple ip addresses (e.g. performance distribution is not an issue here) and therefore we explicitly define a single LB for a single purpose. If you want to have the same LB semantics but with a different LB IP, you can have this by defining a copy of the LB object, but you can also realise as a customer that this may not be intended